runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/work/refresh/FixedStreamHeartbeatSender.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.dataflow.worker.windmill.work.refresh;
+
+import java.util.Objects;
+import org.apache.beam.runners.dataflow.worker.streaming.RefreshableWork;
+import org.apache.beam.runners.dataflow.worker.windmill.client.WindmillStream.GetDataStream;
+import org.apache.beam.sdk.annotations.Internal;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * {@link HeartbeatSender} implementation that sends heartbeats directly on the underlying stream if
+ * the stream is not closed.
+ *
+ * @implNote
+ *     <p>{@link #equals(Object)} and {@link #hashCode()} implementations delegate to internal
+ *     {@link GetDataStream} implementations so that requests can be grouped and sent on the same
+ *     stream instance.
+ *     <p>This class is a stateless decorator to the underlying stream.
+ */
+@Internal
+public final class FixedStreamHeartbeatSender implements HeartbeatSender {
+  private static final Logger LOG = LoggerFactory.getLogger(FixedStreamHeartbeatSender.class);
+  private final GetDataStream getDataStream;
+
+  private FixedStreamHeartbeatSender(GetDataStream getDataStream) {
+    this.getDataStream = getDataStream;
+  }
+
+  public static FixedStreamHeartbeatSender create(GetDataStream getDataStream) {
+    return new FixedStreamHeartbeatSender(getDataStream);
+  }
+
+  @Override
+  public void sendHeartbeats(Heartbeats heartbeats) {
+    if (getDataStream.isShutdown()) {
```

Comment:
instead of isShutdown check shoudl it be try/catch around getDataStream.refreshActiveWork?
That would also log and fail the work items in the case we started refreshing but then stream was shutdown.

Reply:
done